File: trab3/python/falsaposicao.py
from base import *


@algoritmo
def falsaposicao(self, f, xi, xf):
    if f(xi) * f(xf) > 0:
        raise ErroIntervalo
    for _ in xrange(self.nmi):
        p = f(xi) / f(xf)
        xm = xi - (xf - xi) / (1 / p - 1)
        if (abs(xf - xm) < self.tol or abs(xi - xm) < self.tol) and abs(f(xm)) < self.tol:
            return xm
        (xi, xf) = (xi, xm) if f(xm) * f(xi) < 0 else (xm, xf)
    raise ErroNMI

File: trab3/ruby/falsaposicao.rb
# encoding: utf-8
# Algoritmo BISSEÇÃO
require "./base"

Calcnum::algoritmo :falsaposicao do |xi, xf|
  break :erro_intervalo if f(xi) * f(xf) > 0
  tabela 'k', 'xi', 'xf', 'xm', 'f(xm)'
  iterar do |k|
    p = f(xi) / f(xf)
    xm = xi - (xf - xi) * p / (1 - p)
    fxm = f(xm)
    tabela k, xi, xf, xm, fxm
    break xm if ((xf - xm).abs < tol or (xi - xm).abs < tol) and fxm.abs < tol
    xi, xf = f(xm) * f(xi) < 0 ? [xi, xm] : [xm, xf]
  end
end

